src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -514,27 +478,79 @@ else if (null != mHeader) {
      * @param mateReferenceIndex Must either equal NO_ALIGNMENT_REFERENCE_INDEX (-1) indicating no reference, or the
      *                       record must have a SAMFileHeader and the index must exist in the associated sequence
      *                       dictionary.
-     * @throws IllegalStateException if the SAMFileHeader is null for this record or the mate reference index is not
-     *         found in the sequence dictionary for this record.
+     * @throws IllegalStateException if the SAMFileHeader is null for this record
+     * @throws IllegalArgumentException if the mate reference index is not found in the sequence dictionary in the header for this record.
      */
     public void setMateReferenceIndex(final int mateReferenceIndex) {
-        if (mateReferenceIndex == NO_ALIGNMENT_REFERENCE_INDEX) {
-            mMateReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            mMateReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-        }
-        else if (null == mHeader) {
-            throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the mate reference index");
+        // resolveNameFromIndex throws if the index can't be resolved
+        setMateReferenceName(resolveNameFromIndex(mateReferenceIndex, mHeader));
+        mMateReferenceIndex = mateReferenceIndex;
+    }
+
+    /**
+     * Static method that resolves and returns the reference index corresponding to a given reference name.
+     *
+     * @param referenceName If {@code referenceName} is NO_ALIGNMENT_REFERENCE_NAME, the value NO_ALIGNMENT_REFERENCE_INDEX
+     *                      is returned directly. Otherwise {@code referenceName}  must be looked up in the header's sequence
+     *                      dictionary.
+     * @param header SAMFileHeader to use when resolving {@code referenceName} to an index. Must be non null if the
+     *                {@code referenceName} is not NO_ALIGNMENT_REFERENCE_NAME.
+     * @param strict if true, throws if {@code referenceName}  does not appear in the header's sequence dictionary
+     * @returns the reference index corresponding to the {@code referenceName}, or null if strict is false and {@code referenceName}
+     * does not appear in the header's sequence dictionary.
+     * @throws IllegalStateException if {@code referenceName} is not equal to NO_ALIGNMENT_REFERENCE_NAME and the header is null
+     * @throws IllegalArgumentException if strict is true and the name does not appear in header's sequence dictionary.
+     *
+     * Does not mutate the SAMRecord.
+     */
+    protected static Integer resolveIndexFromName(final String referenceName, final SAMFileHeader header, final boolean strict) {
```

Comment:
Please add direct unit tests for both `resolveIndexFromName()` and `resolveNameFromIndex()` that exercise all the possible cases.